3단계 질문 삭제하기 리팩터링 리뷰 요청드립니다

yevgnenll opened the pull request:
용근님 안녕하세요!!

지난번 리뷰 덕분에 여러 깨달음이 있었습니다 다시한번 감사드립니다 :)))))

다음의 생각을 하면서 코드를 작성 하였습니다.

- 질문과 답변은 모두 질문 중심이라는 결론을 얻었습니다
    - 질문은 질문에서 작성됩니다
    - 답변은 답변 독립적으로 작성되는 것이 아닌 질문에서 작성됩니다(고로 `toQuestion`은 필요 없는것 같습니다)
- 연관관계를 불러오는 변수는 `toString()` 에서 제거하였습니다.
- deleteHistory 조차도 domain 의 로직이여야 하지 않을까? 라는 생각이 들었습니다
    - `question.convertDelete()` 의 결과로 `List<DeleteHistory>` 를 생성하도록 하였습니다.
- 따라서 service layer 의 로직은 매우 단순해졌습니다
- 관련하여 테스트 코드 수정이 동반되었습니다

정말 회사 코드를 보여드리고 싶어요
제 회사 코드가 얼마나 달라졌는지, 얼마나 TDD가 반영 되었는지 회사 코드에 보안유지 서약이 아니었다면 벌써 공개 했을거예요 ㅎㅎ

Commit:
refactor: unchecked exception 으로 변경한다

compile 시점에 체크하는 exception 이 아니므로
이 exception 의 용도롤 변경한다

Commit:
feat: question 삭제 확인 기능 개선

question 삭제시 작성자와 로그인 유저가 일치하는지 확인을
service 에서 확인하여 exception 을 실행하는 것이 아니라
domain 에서 수행하도록 변경한다

Commit:
feat: CannotDeleteException 이 기본 message를 갖도록 한다

사용하는 측에서 별도의 message 를 입력하지 않으면
기본 message 를 출력하도록 개선한다

Commit:
feat: answer 작성자 로직를 domain 으로 이관

service layer 에서 exception 을 일으키는 것이 아니라
이 책임을 도메인이 수행하도록 변경한다

Commit:
docs: 요구사항 추가

yevgnenll committed:
feat: answer 삭제 기능 구형

answer 내에서 삭제를 수행하는 기능을 구현한다
flag 를 true 로 만드는 것 뿐 아니라 DeleteHistory 생성까지
함께 수행하도록 개선한다

yevgnenll committed:
test: 변경된 테스트 케이스 반영

yevgnenll committed:
feat: question 에서 answer 삭제 기능 구현

service 에서 answer 를 삭제하는 기능을 domain 으로 캡슐화 한다

yevgnenll committed:
test: service 테스트 변경된 로직 반영

불필요한 로직 제거
answer 는 question 에 의해 select 되므로 answer 를 select 하는
코드는 제거하여 불필요한 테스트 코드를 제거한다

yevgnenll committed:
feat: 질문 삭제기능 구현

service layer 에서 수행했던 로직을 domain 으로 흡수한다

yevgnenll committed:
refactor: toString 의 연관관계 제거

toString 에서 굳이 연관관계를 일으킬 이유가 없어 제거한다

yevgnenll committed:
test: 확인을 위해 사용한 sout 제거

yevgnenll committed:
refactor: 불필요한 함수 제거

질문에 대한 답변은 질문에 다는 것이지 답변을 독립적으로 작성하는 것이 아니다
따라서 toQuestion 이라는 함수는 필요 없다.

yevgnenll committed:
refactor: 사상 불일치 해소를 위한 함수 추가

질문 작성, 답변 작성 모두 각 객체에서 일어나는 일이며, 질문 중심으로 일어나는 사건이다
따라서 DB에 의존적인 판단이 아니라 질문 도메인 중심적인 이벤트가 일어나야 한다고 여겨
답변을 다는 행위도 Question 에서 일어나도록 변경한다

yevgnenll committed:
refactor: client 에 제공할 writeAnswer 추가

kingbbode commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -62,17 +63,10 @@ public Answer(Long id, User writer, Question question, String contents) {
+    public void isOwnerOrThrows(User writer) {
+        if (!this.writer.equals(writer)) {
+            throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
         }
-
-        this.question = question;
-        this.question.addAnswer(this);
     }
```
1. 메서드가 여러가지 일을 하고 있는 것 같네요. 한가지 일만 하도록 작성하는 것이 좋습니다.

2. `isOwnerOrThrows` 에  삭제의 의도는 없는 것 같아요. 메시지가 어울리지 않아보입니다. 

두 가지 문제를 종합하였을 때 너무 많은 것을 객체가 책임지게 하려고 했던 것으로 보여요. 객체는 객체가 가져야할 책임에 집중하고 비지니스는 비지니스를 구현해야할 객체가 책임지도록 변경 부탁드립니다.

yevgnenll replied:
아.. 그렇네요 더 좁은 책임을 갖도록 변경하겠습니다

kingbbode commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -99,18 +93,19 @@ public boolean isDeleted() {
+    public DeleteHistory convertDelete() {
+        this.deleted = true;
+        return new DeleteHistory(ContentType.ANSWER, id, writer);
+    }
```
모든 객체에는 관계가 있습니다. 우리가 가장 신경써야할 것은 이 관계를 어느 선에서 끊을 것이고, 어느 방향으로 설정할지일텐데요.

`DeleteHistory` 가 `Answer`, `Question` 이 가져야할 주요 관심사는 아닌 것으로 보여요.  `DeleteHistory` 가 변경될 때 영향을 받게 되는 객체가 무엇이 있는지 생각해보시고 현재의 관계 방향이 옳바르게 설정된 것인지 고려해보는게 좋을 것 같습니다.

yevgnenll replied:
역시 `Question` 과 `DeleteHistory`, `Answer` 과 `DeleteHistory` 가 관계를 갖는건 좀 이상하군요 ㅠ
수정해두도록 하겠습니다

kingbbode commented on src/test/java/qna/domain/AnswerDirectionTest.java:
```diff
@@ -24,14 +24,24 @@
         questionRepository.save(question);
-        assertThat(answerRepository.count()).isEqualTo(9);
+        assertThat(answerRepository.findByQuestionIdAndDeletedFalse(5L).size()).isEqualTo(9);
+        System.out.println("end !!!");
```
ㅎㅎ sout 은 제거 부탁드려요.

yevgnenll replied:
넵 제거 완료했습니다

kingbbode reviewed (changes requested):
코멘트 작성해두었습니다 :)
확인 부탁드려요.

yevgnenll committed:
refactor: 함수의 기능을 좁힌다

로그인 유저가 작성자인지 아닌지만 판단하는 기능을 갖도록
기능의 범위를 좁힌다

yevgnenll committed:
refactor: 객체의 관계 변경

Question 과 DeleteHistory, Answer 과 DeleteHistory 의
관계를 끊고, 이들의 관계는 비즈니스 로직에서 연결하도록 변경한다

yevgnenll committed:
refactor: 불필요한 sout 제거

kingbbode commented on src/main/java/qna/domain/Question.java:
```diff
@@ -133,21 +131,27 @@ public String toString() {
+            result = answer.isOwner(loginUser) && result;
         }
+        return result;
     }
```
`stream` 으로 조금 더 가독이 좋고 간결하게 작성해볼 수 있지 않았을까 하네요 :)

yevgnenll replied:
이 부분 stream 을 사용해서 수정했습니다 ㅎㅎ

kingbbode reviewed (approved):
잘 반영해주셨네요 :)
수고 하셨습니다. 머지하겠습니다!
